File: 090/098.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"math"
	"sort"
	"strconv"
	"strings"
	"sync"
)

const (
	numAnagramRoutines = 4
)

type Anagram struct {
	word1 string
	word2 string
}

type Mask []string

func NewMask(s string) Mask {
	return strings.Split(s, "")
}
func (m Mask) GetMask() string {
	return strings.Join(m, "")
}

type Result struct {
	words    Anagram
	replaced Anagram
}

func loadWords(filename string, output chan string) {
	defer close(output)
	file, err := ioutil.ReadFile("words.txt")
	if err != nil {
		return
	}
	data := bytes.Split(file, []byte{'\n'})
	for d := range data {
		output <- string(data[d])
	}
}

func getAnagrams(words chan string, output chan Anagram) {
	data := make(map[string][]string)

	for word := range words {
		word_parts := strings.Split(word, "")
		sort.Strings(word_parts)
		key := strings.Join(word_parts, "")

		for other_word := range data[key] {
			output <- Anagram{word, data[key][other_word]}
		}
		data[key] = append(data[key], word)
	}
	close(output)
}

func resultWatcher() chan Result {
	result := make(chan Result)
	highest_found := 0
	go func() {
		for r := range result {
			i, _ := strconv.Atoi(r.replaced.word1)
			j, _ := strconv.Atoi(r.replaced.word2)
			if i > highest_found {
				highest_found = i
				fmt.Println("New Hightest Found:", i, r)
			}
			if j > highest_found {
				highest_found = j
				fmt.Println("New Hightest Found:", j, r)
			}
		}
	}()
	return result
}

func anagramFinder(anagrams chan Anagram, results chan Result, wg *sync.WaitGroup) {
	for anagram := range anagrams {
		findPowerPairs(anagram, results)
	}
	wg.Done()
}

func findPowerPairs(anagram Anagram, results chan Result) {
	masks := generatePowers(len(anagram.word1))

	for m := range masks {
		mask := NewMask(strconv.Itoa(m))
		replaced := replace(anagram, mask)
		j, _ := strconv.Atoi(replaced.word2)
		if isSquare(j) && strings.Split(replaced.word2, "")[0] != "0" {
			results <- Result{anagram, replaced}
		}
	}
}

func generatePowers(length int) chan int {
	output := make(chan int)
	go func() {
		start := int(math.Ceil(math.Sqrt(math.Pow(10.0, float64(length-1)))))
		end := int(math.Sqrt(math.Pow(10.0, float64(length)) - 1))
		for i := start; i <= end; i++ {
			output <- i * i
		}
		close(output)
	}()
	return output
}

func isSquare(i int) bool {
	return math.Pow(math.Floor(math.Sqrt(float64(i))), 2) == float64(i)
}

func replace(anagram Anagram, mask Mask) Anagram {
	word_array := strings.Split(anagram.word1, "")
	word_array_2 := strings.Split(anagram.word2, "")
	past_replacements := make(map[string]string)
	mask_assingments := make(map[string]string)

	for i := 0; i < len(mask); i++ {
		this_letter := word_array[i]
		this_mask := mask[i]
		if past_replacements[this_letter] != "" && past_replacements[this_letter] != mask[i] {
			return Anagram{"0", "0"}
		}
		if mask_assingments[this_mask] != "" && mask_assingments[this_mask] != this_letter {
			return Anagram{"0", "0"}
		}
		past_replacements[this_letter] = mask[i]
		mask_assingments[mask[i]] = this_letter
		word_array[i] = mask[i]
	}

	for i := 0; i < len(word_array_2); i++ {
		word_array_2[i] = past_replacements[word_array_2[i]]
	}

	return Anagram{strings.Join(word_array, ""), strings.Join(word_array_2, "")}
}

func main() {
	words := make(chan string)
	go loadWords("words.txt", words)

	anagrams := make(chan Anagram)
	go getAnagrams(words, anagrams)

	var wg sync.WaitGroup

	results := resultWatcher()

	for i := 0; i < numAnagramRoutines; i++ {
		wg.Add(1)
		go anagramFinder(anagrams, results, &wg)
	}
	wg.Wait()
}

File: 090/098.py
#! /usr/bin/env python

from collections import defaultdict
from math import sqrt, ceil

def getWords():
  f = open("words.txt", "rd.txt")
  for line in f.readlines():
    yield line.strip()
  f.close()

def getAnagrams(words):
  m = defaultdict(list)
  for word in words:
    sort = ''.join(sorted(word))
    for word2 in m[sort]:
      yield (word, word2)
    m[sort].append(word)

def getPowers(length):
  start = int(ceil(sqrt(10 ** (length - 1))))
  end = int(sqrt(10**length))
  for i in range(start, end):
    yield i**2


def replace(word1, word2, mask):
  mask = str(mask)
  word_map = dict()
  mask_map = dict()
  new_word1 = [0]*len(word1)
  new_word2 = [0]*len(word1)

  for i in range(len(mask)):
    this_letter = word1[i]
    this_mask = mask[i]
    if this_letter in word_map and word_map[this_letter] != this_mask:
      return ('0', '0')
    if this_mask in mask_map and mask_map[this_mask] != this_letter:
      return ('0', '0')
    mask_map[this_mask] = this_letter
    word_map[this_letter] = this_mask
    new_word1[i] = this_mask

  for i in range(len(word2)):
    new_word2[i] = word_map[word2[i]]

  return (''.join(new_word1), ''.join(new_word2))

def findPowersForAnagram(anagram):
  length = len(anagram[0])

  for power in getPowers(length):
    replaced = replace(anagram[0], anagram[1], power)
    if replaced[1][0] != '0' and isSquare(int(replaced[1])):
      yield (anagram, replaced)

def isSquare(i):
  return int(sqrt(i))**2 == i

def main():
  anagrams = getAnagrams(getWords())

  m = 0
  for anagram in anagrams:
    for power in findPowersForAnagram(anagram):
      t = int(max(power[1]))
      if t > m:
        print power
        m = t

if __name__ == '__main__':
  main();

